### PR TITLE
Add document with suggestions for ID fields

### PR DESCRIPTION
**Pull request by DrewML**
[Rendered Document](/magento/architecture/blob/id-fields/design-documents/graph-ql/id-fields-object-types.md)

### CODE REVIEW

File: design-documents/graph-ql/id-fields-object-types.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    - Most GraphQL client libs automatically use this field for caching. Any other field name will require manual caching logic on the client
+        - https://www.apollographql.com/docs/react/caching/cache-configuration/#assigning-unique-identifiers
+        - https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#key-generation
+- Include an ID field anytime an Object Type _could_ have an ID field
+
+### Do Not
+- Include info in the client-facing description describing how the field is encoded/decoded (should be opaque)
+    - Example: The client shouldn't know if an `ID` is a base64-encoded integer
+- Use `String` or `Int` type
+
```
Comment (tjwiebell) [inline design-documents/graph-ql/id-fields-object-types.md:20]: It feels odd to include this, but SelectedConfigurableOption violates this rule, and causes odd behavior with automatic caching. This `id` field is based on the `ConfigurableProductOption.id`, which is not unique across all responses. Doesn't hurt to be explicit, but consider this optional.
```suggestion
- Use an ID value you cannot guarantee is unique across all responses
```

Comment (DrewML) [reply design-documents/graph-ql/id-fields-object-types.md:20]: > This id field is based on ConfigurableProductOption.id, which is not unique across all responses.

Whoa. Just to make sure I'm reading this right, you're saying that I can get 2 different `ConfigurableProductOption` types from the API that collide on the `id` field, but hold different data?

Comment (tjwiebell) [reply design-documents/graph-ql/id-fields-object-types.md:20]: Screen Shot 2020-07-08 at 4 31 47 PM

Comment (DrewML) [reply design-documents/graph-ql/id-fields-object-types.md:20]: > which I didn't consider might also not be unique 

Oh noooooo. Well, you've definitely highlighted that we've got some work to do, but trending in the right direction!

Going to push an update to the document

Comment (DrewML) [reply design-documents/graph-ql/id-fields-object-types.md:21]: Added this under `Do Not`, let me know if you have suggestions to tweak the wording:

> Use duplicate IDs across a concrete type. In other words, if the client wants to produce a cache key, the concetenation of a `__typename` + `id` field should _always_ be unique.

**Comment by DrewML** [comment]
Added the following under `Do` based on feedback from @prabhuram93:

> Make the `ID` field non-nullable (syntax: `ID!`)

**Comment by DrewML** [comment]
Never have we reached consensus this easily 😂

**Pull request merged**